influxdb3_write/src/write_buffer/buffer_segment.rs

```diff
@@ -0,0 +1,656 @@
+    fn add_table_batches(&mut self, table_batches: HashMap<String, TableBatch>) {
+        for (table_name, table_batch) in table_batches {
+            let write_table_batch = self.table_batches.entry(table_name).or_default();
+
+            for (partition_key, partition_batch) in table_batch.partition_batches {
+                let write_partition_batch = write_table_batch
+                    .partition_batches
+                    .entry(partition_key)
+                    .or_default();
+                write_partition_batch.rows.extend(partition_batch.rows);
+            }
+        }
+    }
+}
+
+pub struct BufferedWrite {
+    pub wal_op: WalOp,
+    pub database_write: DatabaseWrite,
+    pub response_tx: oneshot::Sender<BufferedWriteResult>,
+}
+
+pub struct DatabaseWrite {
+    pub(crate) db_name: NamespaceName<'static>,
+    pub(crate) table_batches: HashMap<String, TableBatch>,
+}
+
+impl DatabaseWrite {
+    pub fn new(
+        db_name: NamespaceName<'static>,
+        table_batches: HashMap<String, TableBatch>,
+    ) -> Self {
+        Self {
+            db_name,
+            table_batches,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct WriteSummary {
+    pub segment_id: SegmentId,
+    pub sequence_number: SequenceNumber,
+    pub buffer_size: usize,
+}
+
+#[derive(Debug, Default)]
+struct DatabaseBuffer {
+    table_buffers: HashMap<String, TableBuffer>,
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct TableBuffer {
+    pub partition_buffers: HashMap<String, PartitionBuffer>,
+}
+
+impl TableBuffer {
+    #[allow(dead_code)]
+    pub fn partition_buffer(&self, partition_key: &str) -> Option<&PartitionBuffer> {
+        self.partition_buffers.get(partition_key)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct PartitionBuffer {
+    rows: Vec<Row>,
+    timestamp_min: i64,
+    timestamp_max: i64,
+}
+
+impl Default for PartitionBuffer {
+    fn default() -> Self {
+        Self {
+            rows: Vec::new(),
+            timestamp_min: i64::MAX,
+            timestamp_max: i64::MIN,
+        }
+    }
+}
+
+impl PartitionBuffer {
+    pub fn add_rows(&mut self, rows: Vec<Row>) {
+        for row in rows {
+            self.timestamp_min = self.timestamp_min.min(row.time);
+            self.timestamp_max = self.timestamp_max.max(row.time);
+            self.rows.push(row);
+        }
+    }
+
+    pub fn timestamp_min_max(&self) -> TimestampMinMax {
+        TimestampMinMax {
+            min: self.timestamp_min,
+            max: self.timestamp_max,
+        }
+    }
+
+    pub fn row_count(&self) -> usize {
+        self.rows.len()
+    }
+
+    pub fn rows_to_record_batch(
+        &self,
+        schema: &Schema,
+        column_types: &BTreeMap<String, i16>,
+    ) -> RecordBatch {
+        let row_count = self.rows.len();
+        let mut columns = BTreeMap::new();
+        for (name, column_type) in column_types {
+            match ColumnType::try_from(*column_type).unwrap() {
```

Comment:
Along with this one, there are a few `unwrap`s (and a `panic`) in this function. It looks like the function that calls this is fallible (called on 192):
https://github.com/influxdata/influxdb/blob/4d9095e58d2482ecd6874d1d68ef72025d4dc849/influxdb3_write/src/write_buffer/mod.rs#L174-L192
So, `rows_to_record_batch` could also be fallible. Any opposition to using [`anyhow`](https://docs.rs/anyhow/latest/anyhow/) to simplify errors for functions like this?

Reply:
Might be better for a follow on, I think it is more or less switching from
```rust
rows_to_record_batch(/* ... */) -> RecordBatch {
    /* ... */
}
```
to
```rust
rows_to_record_batch(/* ... */) -> Result<RecordBatch, anyhow::Error> {
    /* ... */
}
```
and using [`anyhow::Context`](https://docs.rs/anyhow/latest/anyhow/trait.Context.html) in place of `unwrap`/`panic`/`expect` here. However, this "poisons" the calling function with `anyhow`, which then needs to handle it explicitly or propagate it up, etc.

Once you've wired it up, the error handling could be addressed more holistically in a follow on.

Reply:
I fixed up the `get_table_chunks` to return a `DataFusionError` rather than unwrap. I'll address rows_to_record_batch later when I rework the buffer